Shared Azure middleware between Azure Monitor and Prometheus datasources

kostrse opened the pull request:
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

The goal is to move Azure authentication implementation into a separated library ([grafana/grafana-azure-sdk-go](https://github.com/grafana/grafana-azure-sdk-go)). For this purpose, this PR does preparation and work switches Azure Monitor and Prometheus datasource onto the same middleware and defines common contract.

What changed:

1. Azure Monitor and Prometheus use the same Azure middleware preregistered in http client provider by default.
2. Datasources only needs to set credentials and scopes into `sdkhttpclient.Options` to enable Azure authentication.
3. All Azure datasources should pass **scopes** to Azure middleware as a common contract, not Resource ID (audience). 

Related #33809

**Special notes for your reviewer**:

Prometheus datasource still stores Resource ID (audience) in JsonData, but it's implementation detail of the Prometheus datasource. In near future scopes for Prometheus will be preconfigured based on Azure cloud, so user won't need to set them in JsonData.

Azure configuration directly via JsonData stays behind the experimental feature flag `httpclientprovider_azure_auth` (preserved the original name of the feature flag to avoid breaking changes). Before this feature flag was controlling middleware in `httpclientprovider`, now the middleware is always enabled. So, effectively behavior is the same.

In future, I'm considering removing this possibility of direct configuration of Azure authentication via JsonData (_datasource_service.go_), It's currently needed by templating queries in Prometheus datasource (see #37784), so it stays under experimental feature flag for now.

Commit:
Scopes in Azure middleware

Commit:
Use common Azure middleware in Azure Monitor

Commit:
Enable Azure middleware without feature flag

Commit:
Apply feature flag to JsonData configuration of Azure auth

Commit:
Enforce feature flag in Prometheus datasource

Commit:
Prometheus provider tests

Commit:
Datasource service tests

Commit:
Fix http client provider tests

Commit:
Pass sdkhttpclient.Options by reference

marefr commented on pkg/infra/httpclient/httpclientprovider/http_client_provider.go:
```diff
@@ -27,6 +26,7 @@ func New(cfg *setting.Cfg, tracer tracing.Tracer, features featuremgmt.FeatureTo
 		SetUserAgentMiddleware(userAgent),
 		sdkhttpclient.BasicAuthenticationMiddleware(),
 		sdkhttpclient.CustomHeadersMiddleware(),
+		AzureMiddleware(cfg),
```
It's not optimal that every single outgoing datasource HTTP request will pass through the azure middleware which will look if the CustomOptions contains a certain key, i.e. no strong typing. For example [this](https://github.com/grafana/grafana/blob/2e6f14d17f832cdeddbbc4f63fa0f388f877a6d9/pkg/infra/httpclient/httpclientprovider/sigv4_middleware.go#L19-L21) makes things much cleaner and easier to reason about. My [comment](https://github.com/grafana/grafana-plugin-sdk-go/pull/468#pullrequestreview-897329203) in your other PR is highly relevant.

There's no feature toggle or setting to enable/disable azure auth since it's being used by builtin azure datasources, is that correctly understood?

kostrse replied:
Not for every request, for every instantiation of `http.Client`, i.e. once per each datasource, given the client instance is preserved.

`AzureMiddleware` already has short return:
https://github.com/kostrse/grafana/blob/09da311c04ab69b0f94fe5fb51ef605adf291826/pkg/infra/httpclient/httpclientprovider/azure_middleware.go#L21
It excludes the middleware out of request/response round trip.

Once strongly-typed schema introduced to `httpclient.Options`, it will eliminate overhead of working with `CustomOptions`:

```go
func AzureMiddleware(cfg *setting.Cfg) httpclient.Middleware {
	return httpclient.NamedMiddlewareFunc(azureMiddlewareName, func(opts httpclient.Options, next http.RoundTripper) http.RoundTripper {
		if opts.Azure == nil || opts.Azure.Credentials == nil {
			return next
		}

		...

		return aztokenprovider.ApplyAuth(tokenProvider, opts.Azure.Scopes, next)
	})
}
```

I removed the feature toggle because, yes, the middleware will be used by Azure Monitor and Prometheus and it should work out of box, without any feature flags.

Since `infra/httpclient/httpclientprovider` is available only to internal plugins, external plugins will still need to manually register the middleware for themselves. Do we have a way to enable it by default for external plugins too?
Considering this, does it make sense to require all plugins (internal and external) to enable it explicitly for consistency?

marefr commented on pkg/services/datasources/service/datasource_service.go:
```diff
@@ -303,7 +305,8 @@ func (s *Service) httpClientOptions(ds *models.DataSource) (*sdkhttpclient.Optio
 
-	if ds.JsonData != nil {
+	// Azure authentication
+	if ds.JsonData != nil && s.features.IsEnabled(featuremgmt.FlagHttpclientproviderAzureAuth) {
```
Here you use a feature toggle and not in http_client_provider.go - can you explain why?

kostrse replied:
I would want to completely remove ability to enable Azure authentication directly from `JsonData` and I would want to require all datasources to configure it programmatically via `opts.Azure`.

This code block behind the feature toggle still exists only because templating queries still require it.

More context here:
https://github.com/grafana/grafana-plugin-sdk-go/pull/468#issuecomment-1057622710

Also discussed here:
https://github.com/grafana/grafana/issues/37784#issuecomment-1030449903

sunker reviewed (commented):
Moving the token provider logic from the data source to the http client provider seem like the right thing to do to me, but it will make it harder for the data source to add custom logic. For example, the ADX data source recently added support for OBO token exchange. When this feature is enabled, we don't want the http client to add yet another token. Anyway, I think there should be ways to solve that. Let's discuss this further tomorrow!

kostrse commented:
> Moving the token provider logic from the data source to the http client provider seem like the right thing to do to me, but it will make it harder for the data source to add custom logic. For example, the ADX data source recently added support for OBO token exchange. When this feature is enabled, we don't want the http client to add yet another token. Anyway, I think there should be ways to solve that. Let's discuss this further tomorrow!

In case if a datasource wants to register own custom middleware, it can leave `opt.Azure` struct unset, so the default middleware won't be activated.

Since `infra/httpclient/httpclientprovider` is available only to internal plugins, external plugins will still need to manually register the `AzureMiddleware` by themselves.

Whether the `AzureMiddleware` should be registered by default for both internal and external plugins, we need to discuss tomorrow.

Also commented here:
https://github.com/grafana/grafana/pull/46002/files#r818266619

Commit:
Add middleware to httpclient.Options

kostrse commented:
@marefr I updated the PR. I changed setting of CustomOptions/AzureOptions to parametrized middleware.
- [pkg/tsdb/azuremonitor/httpclient.go#L42](https://github.com/kostrse/grafana/blob/9632d41219f1fbdc33a07501eb95b72a6d3dcbc9/pkg/tsdb/azuremonitor/httpclient.go#L42)
- [pkg/tsdb/prometheus/promclient/provider_azure.go#L45](https://github.com/kostrse/grafana/blob/9632d41219f1fbdc33a07501eb95b72a6d3dcbc9/pkg/tsdb/prometheus/promclient/provider_azure.go#L45)
- [pkg/services/datasources/service/datasource_service.go#L335](https://github.com/kostrse/grafana/blob/9632d41219f1fbdc33a07501eb95b72a6d3dcbc9/pkg/services/datasources/service/datasource_service.go#L335)

Implementation of `azhttpclient.AddAzureAuthentication` looks like this:
- [pkg/tsdb/azuremonitor/azhttpclient/options.go#L9](https://github.com/kostrse/grafana/blob/9632d41219f1fbdc33a07501eb95b72a6d3dcbc9/pkg/tsdb/azuremonitor/azhttpclient/options.go#L9)

I actually like this way more than having `AzureOptions` section in `httpclient.Options` for a few reasons:
- No need `httpclient.Options` know about the Azure. We probably don't need to make any changes in Plugin SDK at all.
- No need to register the middleware in `httpclientprovider` by default. Configuration experience is going to be consistent for both internal and external plugins.

Diff of changes since previous iteration:
https://github.com/kostrse/grafana/commit/9632d41219f1fbdc33a07501eb95b72a6d3dcbc9

Please let me know what do you think.

Commit:
Merge branch 'main' into azmon-common-middleware

Commit:
Remove dependency on Grafana settings

marefr reviewed (commented):
In general looks good. Added a nit comment. 

We've discussed approach with how the middleware gets a hold of the credentials. You prefer parsing/getting the credentials before deciding to call/calling the middleware. 

I already provided this feedback elsewhere, but adding it here for anyone else that's interested. I prefer the default parsing/getting the credentials to be hidden inside the azure middleware. it would be much nicer abstraction and less code in Grafana. And as a bonus the middleware can decide whether to enable the middleware or just return "next" middleware. Another bonus would be to simplify for the consumers (plugin developers) not having to think about how to parse/get credentials unless they have custom (authentication) needs.

Commit:
Unit-tests updated

Commit:
Merge branch 'main' into azmon-common-middleware

Commit:
Fix ds_proxy_test

Commit:
Fix service_test

kostrse commented:
Unit-tests updated

sunker reviewed (commented):
Although I agree with Marcus it would be preferable to move the credential and scope logic to the AzureMiddleware, I think that needs to be done later since it requires refactoring in the azure monitor data source and we also don't want to add more changes to this PR.

I've tested this using Azure Monitor and it's working fine. I don't have a way of testing Azure auth in Prometheus, but I assume you have done that? 

I'm ready to approve and merge this PR, but would like confirmation from you @kostrse that it's ready before I do so? Please note that it's not yet decided if there will be a 8.5.0 release of Grafana prior to 9.0.0. If we merge this PR now, code must be in a releasable state.

marefr reviewed (approved):
LGTM
